src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/IntelliJComponentPresenterBase.java

```diff
@@ -56,12 +73,33 @@ private void removeHighlightInternal() {
       overlayPane.remove();
       overlayPane = null;
     }
+    timer.cancel();
   }
 
   protected abstract GenericHighlighter getHighlighter();
 
+  protected abstract boolean tryToShow();
+
+  public boolean isVisible() {
+    var highlighter = getHighlighter();
+    return highlighter != null && highlighter.getComponent().isVisible();
+  }
+
   @Override
   public void setCancelHandler(CancelHandler cancelHandler) {
     this.cancelHandler = cancelHandler;
   }
+
+  private void startTimer() {
+    timer.scheduleAtFixedRate(new TaskRefresher(), REFRESH_INTERVAL, REFRESH_INTERVAL);
+  }
+
+  private class TaskRefresher extends TimerTask {
+    @Override
+    public void run() {
+      if (!isVisible()) {
+        highlight();
```

Comment:
`highlight` calls `highlightInternal` which calls `startTimer` which creates a new `TaskRefresher`. So there will be an increasing number of `TaskRefresher`s...

I suggest that `highlight`, not `highlightInternal`, starts the timer and both `highlight` and `TaskRefresher::run` call `highlightInternal`. Would that make sense to you?